repose-aggregator/components/filters/keystone-v2/src/main/scala/org/openrepose/filters/keystonev2/KeystoneRequestHandler.scala

```diff
@@ -0,0 +1,313 @@
+        serviceClientResponse.getStatus match {
+          case statusCode if statusCode >= 200 && statusCode < 300 =>
+            val jsonResponse = Source.fromInputStream(serviceClientResponse.getData).getLines().mkString("")
+            val json = Json.parse(jsonResponse)
+            Try(Success((json \ "access" \ "token" \ "id").as[String])) match {
+              case Success(s) => s
+              case Failure(f) =>
+                Failure(IdentityCommunicationException("Token not found in identity response during admin authentication", f))
+            }
+          case SC_REQUEST_ENTITY_TOO_LARGE | SC_TOO_MANY_REQUESTS =>
+            Failure(OverLimitException(buildRetryValue(serviceClientResponse), "Rate limited when getting admin token"))
+          case statusCode if statusCode >= 500 =>
+            Failure(IdentityCommunicationException("Identity Service not available to get admin token"))
+          case _ => Failure(new Exception("Unable to successfully get admin token from Identity"))
+        }
+      case Failure(x) => Failure(new Exception("Failure communicating with identity during admin authentication", x))
+    }
+  }
+
+  final def validateToken(validatingToken: String, validatableToken: String): Try[ValidToken] = {
+    def extractUserInformation(keystoneResponse: InputStream): Try[ValidToken] = {
+      val input: String = Source.fromInputStream(keystoneResponse).getLines mkString ""
+      try {
+        val json = Json.parse(input)
+        //Have to convert it to a vector, because List isn't serializeable in 2.10
+        val roleNames = (json \ "access" \ "user" \ "roles" \\ "name").map(_.as[String]).toVector
+        val defaultTenantId = (json \ "access" \ "token" \ "tenant" \ "id").as[String]
+        val tenantIds = (json \ "access" \ "user" \ "roles" \\ "tenantId").map(_.as[String]).toVector
+        val userId = (json \ "access" \ "user" \ "id").as[String]
+        val username = (json \ "access" \ "user" \ "name").as[String]
+        val tenantName = (json \ "access" \ "token" \ "tenant" \ "name").as[String]
+        val defaultRegion = (json \ "access" \ "user" \ "RAX-AUTH:defaultRegion").asOpt[String]
+        val contactId = (json \ "access" \ "user" \ "RAX-AUTH:contactId").asOpt[String]
+        val expirationDate = iso8601ToRfc1123((json \ "access" \ "token" \ "expires").as[String])
+        val impersonatorId = (json \ "access" \ "RAX-AUTH:impersonator" \ "id").asOpt[String]
+        val impersonatorName = (json \ "access" \ "RAX-AUTH:impersonator" \ "name").asOpt[String]
+        val validToken = ValidToken(expirationDate,
+          userId,
+          username,
+          tenantName,
+          defaultTenantId,
+          tenantIds,
+          roleNames,
+          impersonatorId,
+          impersonatorName,
+          defaultRegion,
+          contactId)
+
+        Success(validToken)
+      } catch {
+        case oops@(_: JsResultException | _: JsonProcessingException) =>
```

Comment:
Nope. If the type doesn't match what we're casting it to, we'll get the JsResultException. If the JSON is invalid, we'll get a JsonProcessingException. Both are cases where the contract is being violated.